File: tools/script/compile.jl
#!/usr/bin/julia

# Date : 2022-07-01
# A script to compile Library cgraph in Unix-like and Windows Platforms
# gets source files iteratively from Directory src

PRO = "cgraph"
DIR = "."
INC = joinpath(DIR, "include")
SRC = joinpath(DIR, "src")
SRC_TYPE = joinpath(SRC, "type")
TST = joinpath(DIR, "tests")
LIB = joinpath(DIR, "lib")

CC = "cc"
CFLAGS = split("-std=c89 -Wall -pedantic -fPIC", " ")
CSFLAGS = "-shared"

MODE = "debug"
if MODE == "debug"
  push!(CFLAGS, "-g")
  push!(CFLAGS, "-DDEBUG")
elseif MODE == "release"
  push!(CFLAGS, "-static")
  push!(CFLAGS, "-O2")
end

# package shared library
AR = "ar"
ARFLAGS = "-rcs"

if Sys.iswindows()
  # target files
  LIBSHARED = joinpath(LIB, "lib$PRO.dll")
  LIBSTATIC = joinpath(LIB, "lib$PRO.lib")
  # test files
  TSTFILE = joinpath(TST, "$PRO.c")
  TSTTARGET = joinpath(TST, "$PRO.exe")
else
  # target files
  LIBSHARED = joinpath(LIB, "lib$PRO.so")
  LIBSTATIC = joinpath(LIB, "lib$PRO.a")
  # test files
  TSTFILE = joinpath(TST, "$PRO.c")
  TSTTARGET = joinpath(TST, PRO)
end

# source files
# get all subdirectories
function getsubdirs(path, dirs)
  for item in readdir(path)
    subpath = joinpath(path, item)
    if isdir(subpath) && nothing == match(r"^\.", item)
      push!(dirs, subpath)
      getsubdirs(subpath, dirs)
    end
  end
end

SRCS = []
getsubdirs(SRC, SRCS)

# get all source files from subdirectories
CFILES = []
for dir in SRCS
  for item in readdir(dir)
    subpath = joinpath(dir, item)
    if isfile(subpath) && nothing != match(r"^((?!\.).)*\.c$", item)
      push!(CFILES, subpath)
    end
  end
end

args = ARGS
if length(args) == 0
  mkpath(LIB)
  OFILES = []
  for file in CFILES
    obj = replace(file, r"\.c$"=>".o")
    dep = replace(file, r"\.c$"=>".d")
    println("compile $file to $obj")
    run(`$CC $CFLAGS -I$INC -I$SRC_TYPE -c $file -o $obj -MD -MF $dep`)
    push!(OFILES, obj)
  end
  println("compile $LIBSHARED")
  run(`$CC $CSFLAGS -o $LIBSHARED $OFILES`)
  println("compile $LIBSTATIC")
  run(`$AR $ARFLAGS $LIBSTATIC $OFILES`)
elseif args[1] == "test"
  println("compile $TSTFILE to $TSTTARGET")
  run(`$CC $CFLAGS -I$INC -o $TSTTARGET $TSTFILE -L$LIB -static -l$PRO -lm`)
  tst_path = joinpath(TST, "elements.csv")
  println("test $TSTTARGET with $tst_path")
  run(`$TSTTARGET $tst_path`)
elseif args[1] == "clean"
  for file in CFILES
    obj = replace(file, r"\.c$"=>".o")
    println("clean ", obj)
    rm(obj, force=true)
    dep = replace(file, r"\.c$"=>".d")
    println("clean ", dep)
    rm(dep, force=true)
  end
  println("clean $LIBSTATIC")
  rm(LIBSTATIC, force=true)
  println("clean $LIBSHARED")
  rm(LIBSHARED, force=true)
  println("clean $TSTTARGET")
  rm(TSTTARGET, force=true)
elseif args[1] == "distclean"
  for file in CFILES
    obj = replace(file, r"\.c$"=>".o")
    println("clean ", obj)
    rm(obj, force=true)
    dep = replace(file, r"\.c$"=>".d")
    println("clean ", dep)
    rm(dep, force=true)
  end
  println("clean $LIBSTATIC")
  rm(LIBSTATIC, force=true)
  println("clean $LIBSHARED")
  rm(LIBSHARED, force=true)
  println("clean $LIB")
  rm(LIB, force=true)
  println("clean $TSTTARGET")
  rm(TSTTARGET, force=true)
elseif args[1] == "help"
  println("$PROGRAM_FILE    <target>")
  println("<target>: ")
  println("                    compile cgraph")
  println("          test      test cgraph")
  println("          clean     clean all the generated files")
  println("          distclean clean all the generated files and directories")
  println("          help      commands to this program")
else
  println(args[1], " is an unsupported command")
  println("use \"$PROGRAM_FILE help\" to know all supported commands")
end

File: tools/script/compile.go
#!/usr/bin/env gorun

// Date : 2022-07-01
// A script to compile Library cgraph in Unix-like and Windows Platforms
// gets source files iteratively from Directory src

package main

import (
  "fmt"
  "io/ioutil"
  "os"
  "os/exec"
  "regexp"
  "path"
  "runtime"
  "strings"
)

func getsubdirs(path, dirs) {
  dirs, _ := ioutil.ReadDir(entry)
  for _, item := range dirs {
    matched, _ := regexp.MatchString(`^(?!\.)`, item.Name())
    if matched {
      if item.IsDir() {
        append(dirs, item)
        getsubdirs(item, dirs)
      }
    }
  }
}

func main() {
  var PRO string = "cgraph"
  var DIR string = "."
  var INC string = path.Join(DIR, "include")
  var SRC string = path.Join(DIR, "src")
  var SRC_TYPE string = os.path.join(SRC, "type")
  var TST string = path.Join(DIR, "tests")
  var LIB string = path.Join(DIR, "lib")

  var CC string = "cc"
  var CFLAGS []string = strings.Split("-std=c89 -Wall -pedantic -fPIC", " ")
  var CSFLAGS string = "-shared"

  var MODE string = "debug"
  if MODE == "debug" {
    CFLAGS = append(CFLAGS, "-g", "-DDEBUG")
  } else if MODE == "release" {
    CFLAGS = append(CFLAGS, "-static", "-O2")
  }

  var AR string = "ar"
  var ARFLAGS string = "-rcs"

  // source files
  // get all subdirectories
  var SRCS []string
  getsubdirs(SRC, SRCS)

  // get all source files from subdirectories
  var CFILES []string
  for _, dir := range SRCS {
    items, _ := ioutil.ReadDir(dir)
    for _, item := range items {
      if item.IsFile() {
        matched, _ := regexp.MatchString(`^(?!\.).*?\.c$`, item.Name())
        if matched {
          CFILES = append(CFILES, path.Join(dir, item.Name()))
        }
      }
    }
  }

  var LIBSHARED string 
  var LIBSTATIC string 
  var TSTFILE string 
  var TSTTARGET string
  
  if runtime.GOOS == "windows" {
    LIBSHARED = path.Join(LIB, "lib" + PRO + ".dll")
    LIBSTATIC = path.Join(LIB, "lib" + PRO + ".lib")
    TSTFILE   = path.Join(TST, PRO + ".c")
    TSTTARGET = path.Join(TST, PRO + ".exe")
  } else {
    LIBSHARED = path.Join(LIB, "lib" + PRO + ".so")
    LIBSTATIC = path.Join(LIB, "lib" + PRO + ".a")
    TSTFILE   = path.Join(TST, PRO + ".c")
    TSTTARGET = path.Join(TST, PRO)
  }

  var args []string = os.Args
  if len(args) == 1 {
    if _, err := os.Stat(LIB); os.IsNotExist(err) {
      os.MkdirAll(LIB, 0777)
    }
    var OFILES []string
    for _, file := range CFILES {
      var obj string = regexp.MustCompile(`\.c$`).ReplaceAllString(file, ".o")
      var dep string = regexp.MustCompile(`\.c$`).ReplaceAllString(file, ".d")
      fmt.Printf("compile %s to %s\n", file, obj)
      ARGS := append(CFLAGS, "-I"+INC, "-c", file, "-o", obj, "-MD", "-MF", dep)
      cmd := exec.Command(CC, ARGS...)
      cmd.Run()
      OFILES = append(OFILES, obj)
    }
    fmt.Printf("compile %s\n", LIBSHARED)
    ARGS0 := []string{CSFLAGS, "-o", LIBSHARED}
    ARGS0 = append(ARGS0, OFILES...)
    cmd0 := exec.Command(CC, ARGS0...)
    cmd0.Run()
    fmt.Printf("compile %s\n", LIBSTATIC)
    ARGS1 := []string{ARFLAGS, LIBSTATIC}
    ARGS1 = append(ARGS1, OFILES...)
    cmd1 := exec.Command(AR, ARGS1...)
    cmd1.Run()
  } else if args[1] == "test" {
    fmt.Printf("compile %s to \n", TSTFILE, TSTTARGET)
    ARGS0 := append(CFLAGS, "-I"+INC, "-o", TSTTARGET, TSTFILE, "-L"+LIB, "-static", "-l"+PRO, "-lm")
    cmd0 := exec.Command(CC, ARGS0...)
    cmd0.Run()
    fmt.Printf("test %s with %s/elements.csv", TSTTARGET, TST)
    cmd1 := exec.Command(TSTTARGET, TST, "/elements.csv")
    cmd1.Run()
  } else if args[1] == "clean" {
    for _, file := range CFILES {
      var obj string = regexp.MustCompile(`\.c$`).ReplaceAllString(file, ".o")
      fmt.Printf("clean %s\n", obj)
      os.Remove(obj)
      var dep string = regexp.MustCompile(`\.c$`).ReplaceAllString(file, ".d")
      fmt.Printf("clean %s\n", dep)
      os.Remove(dep)
    }
    fmt.Printf("clean %s\n", LIBSHARED)
    os.Remove(LIBSHARED)
    fmt.Printf("clean %s\n", LIBSTATIC)
    os.Remove(LIBSTATIC)
    fmt.Printf("clean %s\n", TSTTARGET)
    os.Remove(TSTTARGET)
  } else if args[1] == "distclean" {
    for _, file := range CFILES {
      var obj string = regexp.MustCompile(`\.c$`).ReplaceAllString(file, ".o")
      fmt.Printf("clean %s\n", obj)
      os.Remove(obj)
      var dep string = regexp.MustCompile(`\.c$`).ReplaceAllString(file, ".d")
      fmt.Printf("clean %s\n", dep)
      os.Remove(dep)
    }
    fmt.Printf("clean %s\n", LIBSHARED)
    os.Remove(LIBSHARED)
    fmt.Printf("clean %s\n", LIBSTATIC)
    os.Remove(LIBSTATIC)
    fmt.Printf("clean %s\n", LIB)
    os.RemoveAll(LIB)
    fmt.Printf("clean %s\n", TSTTARGET)
    os.Remove(TSTTARGET)
  } else if args[1] == "help" {
    fmt.Printf("%s    <target>\n", args[0])
    fmt.Printf("<target>: \n")
    fmt.Printf("                    compile cgraph\n")
    fmt.Printf("          test      test cgraph\n")
    fmt.Printf("          clean     clean all the generated files\n")
    fmt.Printf("          distclean clean all the generated files and directories\n")
    fmt.Printf("          help      commands to this program\n")
  } else {
    fmt.Printf("%s is an unsupported command\n", args[1])
    fmt.Printf("use \"%s help\" to know all supported commands\n", args[0])
  }
}
